Simplify device manager: make endpoint stateless

figo opened the pull request:
While reviewing devicemanager code, found the caching layer on endpoint is redundant.
Here are the 3 related objects in picture:

**devicemanager  <->  endpoint <-> plugin**

plugin is the source of truth for devices and device health status.
devicemanager maintain healthyDevices, unhealthyDevices, allocatedDevices based on updates
from plugin.

So there is no point for endpoint to cache devices, this patch is removing the cache layer,
endpoint becomes stateless, which i believe should be the case (but i do welcome review
if i missed something here).

also removing the Manager.Devices() since i didn't find any caller of this other than test.

if we need to get all devices from manager in future, it just need to return healthyDevices + unhealthyDevices, so don't have to call endpoint after all.

This patch makes code more readable, data model been simplified.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
this patch simplify the device manager code, make it more maintainable.

**Which issue(s) this PR fixes** *:
this is a refactor of device manager code

**Special notes for your reviewer**:
will need to rebase the code if #58755 get checked-in first.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

/sig node
/cc @jiayingz @RenaudWasTaken @vishh @saad-ali @vikaschoudhary16 @vladimirvivien @anfernee

RenaudWasTaken commented:
/ok-to-test

On Sun, Jul 8, 2018, 04:46 k8s-ci-robot <notifications@github.com> wrote:

> [APPROVALNOTIFIER] This PR is *NOT APPROVED*
>
> This pull-request has been approved by: *figo
> <https://github.com/kubernetes/kubernetes/pull/65948#>*
> To fully approve this pull request, please assign additional approvers.
> We suggest the following additional approver: *vishh*
>
> Assign the PR to them by writing /assign @vishh in a comment when ready.
>
> The full list of commands accepted by this bot can be found here
> <https://go.k8s.io/bot-commands>.
>
> The pull request process is described here
> <https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process>
> Needs approval from an approver in each of these files:
>
>    - *pkg/kubelet/cm/devicemanager/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/devicemanager/OWNERS>*
>
> Approvers can indicate their approval by writing /approve in a comment
> Approvers can cancel approval by writing /approve cancel in a comment
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/65948#issuecomment-403257677>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACZyE35j_fde7CLXDWysAIbmdfPe8zKBks5uEXKPgaJpZM4VGkOR>
> .
>

figo commented:
/test pull-kubernetes-e2e-kops-aws

figo commented:
/test pull-kubernetes-local-e2e-containerized

neolit123 commented:
@kubernetes/sig-node-pr-reviews 

@figo 
does this need a release note and/or docs updates?

figo commented:
/test pull-kubernetes-local-e2e-containerized

figo commented:
@neolit123 thanks, this is something transparent to user, so i guess don't have to add to release note

vikaschoudhary16 commented:
we might need cached devices for resource classes/compute resources. I would suggest to hold on removing cached devices from endpoint for some time.

figo commented:
@vikaschoudhary16  thanks for jump in, could you clarify which resource class/compute resource need endpoint cache?  i can not find any caller of it.

another fact need be considered is:  DeviceManager already cached devices through healthyDevices and unhealthyDevices lists, we can discuss this in a follow up question after you help to clarify the first question, thank you!

vikaschoudhary16 commented:
https://github.com/kubernetes/community/pull/2265/files

figo commented:
@vikaschoudhary16 thanks for the link, i went through the ResourceClass KEP properly, could you help to clarify further on where endpoint cache is needed for ResourceClass?

i see the simplified model can help ResourceClass implementation potentially, thanks

RenaudWasTaken commented:
@vikaschoudhary16 do you mind explaining your logic here?

jiayingz commented:
@vikaschoudhary16 even with device attributes, I guess it is perhaps still simpler to cache device attribute info in manager.go instead of endpoint.go, especially given that our current model is that device attribute changes requires a node drain? Even if we relax that requirement in the future, I think we still expect device attribute changes as a rare event that an endpoint can just make a blind callback with the full list to update device manager. Given that this PR does simplify the logic quite a bit, should we consider to merge it earlier?

vikaschoudhary16 commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -107,74 +106,62 @@ func TestDevicePluginReRegistration(t *testing.T) {
-		require.Equal(t, 2, len(devices[testResourceName]), "Devices are not updated.")
+
+		_, allocatable, _ := m.GetCapacity()
+		resourceAllocatable, _ := allocatable[v1.ResourceName(testResourceName)]
```
how about checking capacity as well?

figo replied:
done, thanks

vikaschoudhary16 commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -107,74 +106,62 @@ func TestDevicePluginReRegistration(t *testing.T) {
-		require.Equal(t, 2, len(devices2[testResourceName]), "Devices shouldn't change.")
+		_, allocatable, _ = m.GetCapacity()
+		resourceAllocatable, _ = allocatable[v1.ResourceName(testResourceName)]
+		require.Equal(t, int64(2), resourceAllocatable.Value(), "Devices shouldn't change.")
```
will it add value if we check capacity as well?

figo replied:
done, thanks

vikaschoudhary16 commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -336,23 +333,10 @@ func (m *ManagerImpl) Stop() error {
 
 	socketPath := filepath.Join(m.socketdir, r.Endpoint)
-	e, err := newEndpointImpl(socketPath, r.ResourceName, existingDevs, m.callback)
+	old := m.endpoints[r.ResourceName]
```
you can move lock from L344 to here and remove the `if` block at L351?

figo replied:
done, thanks

vikaschoudhary16 commented:
@jiayingz 
>  I guess it is perhaps still simpler to cache device attribute info in manager.go instead of endpoint.go

+1

@RenaudWasTaken i was thinking that we might need to cache device details, but as jiaying said, that  can be cached in manager cache as well. Anyways we are also caching options there so it is even more aligned.

@figo just a few nits and suggestions, mostly lgtm. this has really simplified logic of e.run(). Thanks for the change !!

figo commented:
/test pull-kubernetes-local-e2e-containerized

figo commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -107,74 +106,62 @@ func TestDevicePluginReRegistration(t *testing.T) {
-			t.FailNow()
+		case <-ch:
+		case <-time.After(5 * time.Second):
+			t.Fatalf("timeout wait for manager update")
```
done, thanks

figo commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -107,74 +106,62 @@ func TestDevicePluginReRegistration(t *testing.T) {
-		atomic.StoreInt32(&expCallbackCount, 2)
 		p2.Register(socketName, testResourceName, preStartContainerFlag)
-		// Wait for the second callback to be issued.
+
```
i will keep this extra line just to separate the 3 test steps/sections.

RenaudWasTaken commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -132,30 +146,23 @@ func newManagerImpl(socketPath string) (*ManagerImpl, error) {
 	m.writeCheckpoint()
+
+	if m.updateChan != nil {
+		m.updateChan <- new(interface{})
```
This isn't used anywhere.

figo replied:
thanks for review, this channel is added to help testing, test need to make sure callback been processed and then check stats, please refer to L109 at manager_test.go

RenaudWasTaken replied:
Ah I'm not sure that's a pattern we should use then. I'd rather you wrap that function in another callback in the test.
e.g:

```golang
originalCallback := m.callback
m.callback = func(...) ... {
     originalCallback(...)
     m.updateChan <- new(interface{})
}
```

figo replied:
good idea, done, thanks

RenaudWasTaken commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -132,30 +146,23 @@ func newManagerImpl(socketPath string) (*ManagerImpl, error) {
-		m.unhealthyDevices[resourceName] = sets.NewString()
-	}
-	for _, dev := range kept {
+	m.healthyDevices[resourceName] = sets.NewString()
```
We're loosing the ability to react to a device appearing / disappearing or changing state.
That's not a problem since we're not doing anything now but it probably needs some discussion in a separate issue.

figo replied:
thanks, you are right, we are not doing anything now, we can achieve this capability without endpoint cache,  just need to compare the updated devices list with `healthyDevices` + `unhealthyDevices`,  do you think we need to create an issue to track this?

vikaschoudhary16 replied:
@figo +1

RenaudWasTaken reviewed (commented):
Mostly looks good to me

figo commented:
/test pull-kubernetes-local-e2e-containerized

vikaschoudhary16 commented:
/lgtm

frapposelli commented:
/lgtm

figo committed:
While reviewing devicemanager code, found
the caching layer on endpoint is redundant.

Here are the 3 related objects in picture:
devicemanager <-> endpoint <-> plugin

Plugin is the source of truth for devices
and device health status.

devicemanager maintain healthyDevices,
unhealthyDevices, allocatedDevices based on updates
from plugin.

So there is no point for endpoint caching devices,
this patch is removing this caching layer on endpoint,

Also removing the Manager.Devices() since i didn't
find any caller of this other than test, i am adding a
notification channel to facilitate testing,

If we need to get all devices from manager in future,
it just need to return healthyDevices + unhealthyDevices,
we don't have to call endpoint after all.

This patch makes code more readable, data model been simplified.

RenaudWasTaken reviewed (commented):
/lgtm

jiayingz commented:
/lgtm
/approve
